source/grammar/tests/test_grammar.py

```diff
@@ -1,173 +1,232 @@
-
-    def test_declaration(self):
-        """Test classical and quantum declaration."""
-        self.load_and_compare_yaml("declaration.yaml")
-        self.load_and_compare_yaml("complex.yaml")
-
-    def test_pragma(self):
-        """Test pragma directive."""
-        self.load_and_compare_yaml("pragma.yaml")
-
-    def test_expression(self):
-        """Test expressions."""
-        self.load_and_compare_yaml("binary_expr.yaml")
-        self.load_and_compare_yaml("order_of_ops.yaml")
-        self.load_and_compare_yaml("unary_expr.yaml")
-        self.load_and_compare_yaml("built_in_call.yaml")
-        self.load_and_compare_yaml("sub_and_extern_call.yaml")
-
-    def test_assignment(self):
-        """Test assignment statements."""
-        self.load_and_compare_yaml("assignment.yaml")
-
-    def test_branching(self):
-        """Test branching statements."""
-        self.load_and_compare_yaml("branching.yaml")
-        self.load_and_compare_yaml("branch_binop.yaml")
-
-    def test_loop_and_control_directive(self):
-        """Test loop and control directive statements."""
-        self.load_and_compare_yaml("loop.yaml")
-
-    def test_aliasing(self):
-        """Test alias statements."""
-        self.load_and_compare_yaml("alias.yaml")
-
-    def test_examples(self):
-        """Loop through all example files, parse and verify that no errors are raised.
-
-        Examples located at: ``openqasm/examples``.
-        """
-        examples = os.listdir(self.examples_path)
-        for e in examples:
-            example_file = os.path.join(self.examples_path, e)
-            if os.path.isfile(example_file):
-                tree = build_parse_tree(example_file, using_file=True)
+    directory = pathlib.Path(directory).absolute()
+
+    def parameter(root, file):
+        path = pathlib.Path(root) / file
+        return pytest.param(str(path), id=str(path.relative_to(directory)))
+
+    return [
+        parameter(root, file)
+        for root, _, files in os.walk(directory) for file in files
+        if file.endswith(suffix)
+    ]
+
+
+@pytest.mark.parametrize(
+    "filename",
+    find_files(TEST_DIR / "reference", suffix=".yaml"),
+)
+def test_reference_output(filename):
+    """Test that the reference files parse to the exact expected output."""
+    with open(filename, "r") as file:
+        obj = yaml.load(file, Loader=yaml.FullLoader)
+    # Make sure the YAML files have only the correct keys.
+    assert set(obj) == {"reference", "source"}
+    parsed = openqasm_reference_parser.pretty_tree(program=obj["source"])
+    assert parsed == obj["reference"]
+
+
+@pytest.mark.parametrize(
+    "filename",
+    find_files(REPO_DIR / "examples", suffix=".qasm"),
+)
+def test_examples_parse(filename):
+    """Test that the example QASM3 files all parse without error."""
+    openqasm_reference_parser.pretty_tree(file=filename)
+
+
+class TestInvalidProgramsFailToParse:
+    valid_globals = [
```

Comment:
Would it be possible somehow to check these valid/invalid cases as code not tied to Python/pytest? I understand you did it this way to avoid creating `n` files, but this ties these test cases to our test framework?

Ways around this I can think of:
- We could create Yaml files with these cases
- We could create `n` files and just bite the bullet.
- We could use something like llvm-lit + FileCheck, this would add a significant amount of rework, but would enable a more flexible test suite in the longterm. I think this is too much work.
- Leave it as is 😄 .

Reply:
Perhaps I'll separate them out into separate files, but organised into the groups that I've done it here.  Then I'll have the parametrisation routine split the files by line, ignoring comments?

The reason for not having a separate file for every single line is because that would end up thrashing the disk IO, which is slow (though admittedly in our ~15s test suite, not a _huge_ deal), and also because it's hard to glance over what's being tested if there's a million little files.  I agree that it'd be best to separate them out into some easier files, though.